src/compiler/sys.ts

```diff
@@ -298,6 +298,55 @@ namespace ts {
                 };
             }
 
+            function createWatchedFileSet() {
+                const watchedDirectories = createFileMap<FileWatcher>();
+                const watchedFiles = createFileMap<(fileName: string, removed?: boolean) => void>();
+                const currentDirectory = process.cwd();
+
+                return { addFile, removeFile };
+
+                function addFile(fileName: string, callback: (fileName: string, removed?: boolean) => void): WatchedFile {
+                    const path = toPath(fileName, currentDirectory, getCanonicalPath);
+                    const parentDirPath = toPath(ts.getDirectoryPath(fileName), currentDirectory, getCanonicalPath);
+
+                    if (!watchedDirectories.contains(parentDirPath)) {

```

Comment:
For loose files, we use non-recursive watchers for the current folder and upper level folders, only to detect added `tsconfig.json` files. E.g. if the loose file is `C:\Desktop\foo.ts`, folders being watched are `C:\` and `C:\Desktop`. And you are right that for configured project we don't have watchers for subdirs. Only one recursive watcher at the project root is used.

I agree that it is more efficient to re-use the watcher if it is already there, though the implementation requires unification of the `watchDirectory` and `watchFile` apis, which weren't talking to each other. It was done in my recent updates.